extensions/http/src/main/java/org/mule/extension/http/internal/listener/HttpListener.java

```diff
@@ -133,148 +138,122 @@
   @Placement(tab = ADVANCED, group = CONFIGURATION_OVERRIDES)
   private Boolean parseRequest;
 
-  @Parameter
-  @Optional
-  @DisplayName(RESPONSE_SETTINGS)
-  @Placement(group = RESPONSE_SETTINGS)
-  private HttpListenerResponseBuilder responseBuilder;
-
-  @Parameter
-  @Optional
-  @DisplayName(ERROR_RESPONSE_SETTINGS)
-  @Placement(group = ERROR_RESPONSE_SETTINGS)
-  private HttpListenerResponseBuilder errorResponseBuilder;
-
   private MethodRequestMatcher methodRequestMatcher = AcceptsAllMethodsRequestMatcher.instance();
   private HttpThrottlingHeadersMapBuilder httpThrottlingHeadersMapBuilder = new HttpThrottlingHeadersMapBuilder();
   private String[] parsedAllowedMethods;
   private ListenerPath listenerPath;
   private RequestHandlerManager requestHandlerManager;
-  private MuleEventToHttpResponse muleEventToHttpResponse;
+  private HttpResponseFactory responseFactory;
   private List<ErrorType> knownErrors;
-  @Inject
-  private MuleContext muleContext;
 
-  @Override
-  public synchronized void initialise() throws InitialisationException {
-    if (allowedMethods != null) {
-      parsedAllowedMethods = extractAllowedMethods();
-      methodRequestMatcher = new MethodRequestMatcher(parsedAllowedMethods);
+  @OnSuccess
+  public void onSuccess(@Optional @DisplayName(RESPONSE_SETTINGS) @Placement(
+      group = RESPONSE_SETTINGS) @NullSafe HttpListenerSuccessResponseBuilder responseBuilder,
+                        SourceCallbackContext callbackContext)
+      throws Exception {
+
+    HttpResponseContext context = callbackContext.getVariable(RESPONSE_CONTEXT);
+    HttpResponse httpResponse = buildResponse(responseBuilder, context.isSupportStreaming());
```

Comment:
the caller is always the runtime. The extension should never call this method directly. The runtime catches any thrown exceptoins and calls the onError method.

I get your NB point though, but remember that SDK still doesn't support it. We can revisit this when we implement NB so that we can also `not block` when we're sending the response